### ISSUE TITLE
bmv2 Low throughtput 

### ISSUE THREAD

**Issue by adibrastegarnia** [p4lang/behavioral-model#537]
Hi,
I was wondering why I get low throughput (around 30Mbps) and sometimes it drops to 0 when I use bmv2. If I use Mininet with OVS I can get up to 30Gbps on my computer so I don't think something is wrong with my setup. Someone else asked about this issue before but I didn't find a final solution for that. 

Thanks

**Comment by antoninbas** [comment]
Hi @adibrastegarnia,

I'm pretty sure there have been definitive answers on that topic. bmv2 is not meant to be a production software switch and you will definitely not get the kind of throughput you will with OVS, etc., not even close... I will not give a list of software switch alternatives or the trade-offs you have to make for each one, as I have already answered this multiple times in the past.

That being said, 30Mbps is lower than what I would expect to see. Of course it depends on the P4 program so if yours has 50+ match-action tables, this is not such a surprising number.

There is actually a benchmark script in this repo: mininet/stress_test_ipv4.py. I had not run it in a while so I thought your question was a good opportunity for me to do it. I ran the benchmark script on a server running Linux natively with a reasonably good CPU (you will need 4 dedicated cores to get good results). Here are the steps I followed:
- clone bmv2
- `./autogen.sh && ./configure --disable-logging-macros --disable-elogger 'CFLAGS=-O3' 'CXXFLAGS=-O3'`
- `make -j4`
- `cd mininet`
- `sudo python stress_test_ipv4.py`
Usually users who get terrible throughput do not use the correct `configure` flags.

Here are the results I got:
```
Running iperf measurement 1 of 5
887 Mbps
Running iperf measurement 2 of 5
890 Mbps
Running iperf measurement 3 of 5
907 Mbps
Running iperf measurement 4 of 5
907 Mbps
Running iperf measurement 5 of 5
916 Mbps
Median throughput is 907 Mbps
```

The program used for the benchmark (simple_router.p4) has 3 match-tables (one LPM, 2 exact) and one checksum computation. Some things to keep in mind:
- the throughput will decrease linearly with the number of tables
- if you are using an equivalent P4_16 program compiled with p4lang/p4c, it may decrease the throughput (by 2x?) as I have observed that the generated JSON is usually more complex
- when running the benchmark in a VM, I usually get significantly worst results (2-3x?). It may be an issue with my setup, or it may come from ho2 we do packet IO.

BTW, the target throughout for bmv2 is 10,000 Mpps (or 100Mbps with 1,500 byte frames) for reasonable P4 programs (~30 tables).

**Comment by adibrastegarnia** [comment]
Thanks for your response. I found this post before: 
https://github.com/p4lang/behavioral-model/issues/301

It looks like he turned off some of the options on each interface  to get a higher throughput. That is the reason that I asked this question again.   I should test it on a more powerful server and see can I get a higher throughput.

**Comment by antoninbas** [comment]
Closing this for now; please re-open if you cannot get better numbers, but you'll have to provide more information (e.g. P4 program).

**Comment by shahid2636** [comment]
Hi,
I have installed the switch using as follows
./autogen.sh && ./configure --disable-logging-macros --disable-elogger 'CFLAGS=-O3' 'CXXFLAGS=-O3'

I performed the stress test sudo python stress_test_ipv4.py and the results are as follows

Running iperf measurement 1 of 5
912 Mbps
Running iperf measurement 2 of 5
920 Mbps
Running iperf measurement 3 of 5
905 Mbps
Running iperf measurement 4 of 5
944 Mbps
Running iperf measurement 5 of 5
917 Mbps
Median throughput is 917 Mbps

**system configurations** are as follows.
Processor : Intel® Xeon(R) CPU E5-2620 v4 @ 2.10GHz × 16 
Memory: 16G
2 USB 1G interfaces.

I started Switch in the following way
sudo simple_switch -i 3@enx00e04c680065 -i 4@enx00e04c680048 ho.json

my P4 program only have 3 tables. when I try to run iperf from systems connected to port 3 and 4. I only see the bandwidth around 1 Mbits/sec. I have no idea why is this happening. 

Any suggestions?
